File: projeto/frontend/site/vox/src/app/services/dashboard.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, catchError, forkJoin, of } from 'rxjs';

export interface DashboardOverview {
  totalIssues: number;
  approvedIssues: number;
  pendingModerationIssues: number;
  resolvedIssues: number;
  issuesByStatus: Record<string, number>;
  totalProjects: number;
  publishedProjects: number;
  pendingModerationProjects: number;
  projectsByStatus: Record<string, number>;
  totalUsers: number;
  usersByRole: Record<string, number>;
  issueResolutionRate: number;
}

export interface DashboardFilters {
  from: string;
  to: string;
}

export interface DashboardData {
  overview: DashboardOverview | null;
  moderation: Record<string, unknown> | null;
  engagement: Record<string, unknown> | null;
  categories: Record<string, unknown> | null;
  timeline: Record<string, unknown> | null;
  neighborhoods: Array<Record<string, unknown>> | null;
  projectLifecycle: Record<string, unknown> | null;
}

@Injectable({ providedIn: 'root' })
export class DashboardService {
  private readonly apiUrl = 'http://localhost:8080/api/admin/dashboard';

  constructor(private http: HttpClient) {}

  load(filters: DashboardFilters): Observable<DashboardData> {
    const params = new HttpParams().set('from', filters.from).set('to', filters.to);
    const optional = <T>(request: Observable<T>) => request.pipe(catchError(() => of(null)));

    return forkJoin({
      overview: optional(this.http.get<DashboardOverview>(`${this.apiUrl}/overview`, { params })),
      moderation: optional(this.http.get<Record<string, unknown>>(`${this.apiUrl}/moderacao`, { params })),
      engagement: optional(this.http.get<Record<string, unknown>>(`${this.apiUrl}/engajamento`, { params })),
      categories: optional(this.http.get<Record<string, unknown>>(`${this.apiUrl}/categorias`, { params })),
      timeline: optional(this.http.get<Record<string, unknown>>(`${this.apiUrl}/series-temporais`, { params })),
      neighborhoods: optional(this.http.get<Array<Record<string, unknown>>>(`${this.apiUrl}/mapa/bairros`, { params })),
      projectLifecycle: optional(this.http.get<Record<string, unknown>>(`${this.apiUrl}/projetos/ciclo-vida`, { params }))
    });
  }
}

File: projeto/frontend/app/app_vox/lib/services/dashboard_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'api_client.dart';

class DashboardService {
  Future<Map<String, dynamic>> load({
    required String from,
    required String to,
  }) async {
    final results = await Future.wait([
      _entry('overview', 'overview', from, to),
      _entry('moderation', 'moderacao', from, to),
      _entry('engagement', 'engajamento', from, to),
      _entry('categories', 'categorias', from, to),
      _entry(
        'timeline',
        'series-temporais',
        from,
        to,
        extra: {'granularidade': 'month'},
      ),
      _entry('neighborhoods', 'mapa/bairros', from, to),
      _entry(
        'hotspots',
        'mapa/coordenadas',
        from,
        to,
        extra: {'precision': '3'},
      ),
      _entry('projectLifecycle', 'projetos/ciclo-vida', from, to),
    ]);
    return Map<String, dynamic>.fromEntries(results);
  }

  Future<MapEntry<String, dynamic>> _entry(
    String key,
    String endpoint,
    String from,
    String to, {
    Map<String, String> extra = const {},
  }) async {
    try {
      final uri = Uri.parse(
        '${ApiClient.baseUrl}/api/admin/dashboard/$endpoint',
      ).replace(queryParameters: {'from': from, 'to': to, ...extra});
      final response = await http.get(
        uri,
        headers: await ApiClient.authHeaders(),
      );
      ApiClient.checkResponse(response);
      final decoded = jsonDecode(response.body);
      return MapEntry(key, decoded);
    } catch (_) {
      return MapEntry(key, null);
    }
  }
}
